exonum/src/blockchain/mod.rs

```diff
@@ -239,22 +242,35 @@ impl Blockchain {
 
                 fork.checkpoint();
 
-                let r = panic::catch_unwind(panic::AssertUnwindSafe(|| { tx.execute(&mut fork); }));
+                let (r, transaction_status) = {
+                    let mut context = ExecutionContext::new(&mut fork);
+                    let r =
+                        panic::catch_unwind(panic::AssertUnwindSafe(|| tx.execute(&mut context)));
+                    (r, context.into_status())
+                };
 
-                match r {
-                    Ok(..) => fork.commit(),
+                let transaction_status = match r {
+                    Ok(()) => {
+                        fork.commit();
```

Comment:
Do I understand it correctly that a transaction with an explicitly set `Failed` status may persist changes to the storage anyway? A developer may expect for these changes to be discarded, so if this behavior is intended, it should be described *very* carefully.